### ISSUE TITLE
(Bounty) Incorrect handling of generics such as Vector<T> or Dictionary<K, V> 

### ISSUE THREAD

**Issue by jgranick** [HaxeFoundation/intellij-haxe#528]
Although the editor handles generics in Array<T> correctly, iterating over custom types incorrectly reports errors

```haxe
var vector = new Vector<String> ();
vector.push ("Hello");

for (value in vector) {
    
    trace (value.length); // error in IDE
    
}
```

---

Flowplay is interested in paying to support the development of this and other issues:

https://github.com/TiVo/intellij-haxe/issues/529
https://github.com/TiVo/intellij-haxe/issues/530
https://github.com/TiVo/intellij-haxe/issues/531

Thanks everyone :smile:

**Comment by EBatTiVo** [comment]
Another sample: http://try.haxe.org/#628a0

```haxe
class Test {
    static function main() {
		var vector = new Array<String> ();
		vector.push ("Hello");

		for (value in vector) {
	    		trace (value.length); // error in IDE
		}
	}
}
```

**Comment by EricBishton** [comment]
Duplicate of #25, which has some more extensive examples.

**Comment by EricBishton** [comment]
After some research into why Array<String> works, but OpenFls's Vector<T> class and Flixel's FlxSprite<T> class don't, I've come to the preliminary conclusion that the error lies with the plugin's inability to deal with in-line conditional compilation directives.  (That is: #if, #else, #elseif, and #end, appearing within a code line, rather than on lines by themselves.)  Time to revisit this issue, since fixing it will likely fix a host of other issues: see #29, #115, #417, #516.

**Comment by EricBishton** [comment]
So, I proved myself wrong.  The problem lies in the semantic annotator's (and the type resolver's) inability to track the current type through dynamic variables properly.  Rather than extend the semantic annotator, I looked into compiler completion and got that working.  See PR #543 for the fix.

Leaving this open, though, because we're still reporting the errors.

**Comment by EricBishton** [comment]
Fixed by PRs #616, #603, and partially by #593.
